src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -476,92 +245,223 @@ HTTPProxy::_handleCreateSchema(sh::dto::CreateSchemaRequest&& request) {
-HTTPProxy::_handleBeginTxn(sh::dto::BeginTxnRequest&& request){
+seastar::future<std::tuple<sh::Status, shd::TxnBeginResponse>>
+HTTPProxy::_handleTxnBegin(shd::TxnBeginRequest&& request){
     K2LOG_D(log::httpproxy, "Received begin txn request {}", request);
-    k2::K2TxnOptions opts{
-        .deadline= k2::Deadline<>(request.options.timeout),
-        .priority = static_cast<k2::TxnPriority>(request.options.priority),
+    K2TxnOptions opts{
+        .deadline= Deadline<>(request.options.timeout),
+        .priority = static_cast<dto::TxnPriority>(request.options.priority),
         .syncFinalize = request.options.syncFinalize
     };
     return _client.beginTxn(std::move(opts))
         .then([this](auto&& txn) {
-            K2LOG_D(k2::log::httpproxy, "begin txn: {}", txn.mtr());
+            K2LOG_D(log::httpproxy, "begin txn: {}", txn.mtr());
             auto ts = txn.mtr().timestamp;
-            sh::dto::Timestamp shts{.endCount = ts.endCount, .tsoId = ts.tsoId, .startDelta = ts.startDelta};
-            _txns[shts] = std::move(txn);
-            return MakeHTTPResponse<sh::dto::BeginTxnResponse>(sh::Statuses::S201_Created(""), sh::dto::BeginTxnResponse{.timestamp=shts});
+            shd::Timestamp shts{.endCount = ts.endCount, .tsoId = ts.tsoId, .startDelta = ts.startDelta};
+            if (auto it = _txns.find(shts); it != _txns.end()) {
+                return MakeHTTPResponse<shd::TxnBeginResponse>(sh::Statuses::S500_Internal_Server_Error("duplicate transaction ID detected"), shd::TxnBeginResponse{});
+            }
+            else {
+                _txns.insert(it, {shts, ManagedTxn{.handle=std::move(txn), .queries={}}});
+                return MakeHTTPResponse<shd::TxnBeginResponse>(sh::Statuses::S201_Created(""), shd::TxnBeginResponse{.timestamp=shts});
+            }
         });
 }
 
-HTTPProxy::HTTPProxy() : _client(k2::K23SIClientConfig()) {
+seastar::future<std::tuple<sh::Status, shd::WriteResponse>>
+HTTPProxy::_handleWrite(shd::WriteRequest&& request) {
+    K2LOG_D(log::httpproxy, "Received write request {}", request);
+
+    return seastar::do_with(std::move(request), [this] (auto& request) {
+        return _getSchemas(request.collectionName, request.schemaName, request.value.schemaVersion)
+        .then([this, &request](auto&& schemas) mutable {
+            auto& [status, k2Schema, shdSchema] = schemas;
+            if (!status.is2xxOK()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Status{.code=status.code, .message=status.message}, shd::WriteResponse{});
+            }
+            auto it = _txns.find(request.timestamp);
+            if (it == _txns.end()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Statuses::S410_Gone("transaction does not exist"), shd::WriteResponse{});
+            }
+            dto::SKVRecord k2record(request.collectionName, k2Schema);
+            bool haveKeys = !request.isDelete && request.fieldsForPartialUpdate.size() == 0;
```

Comment:
We need the keys even if it is a delete or partial update in order to route. The K2 client will also verify the keys are present in the k2record as part of the write, so we can probably just pass true to the constructor.

Reply:
good point. i'll just pass true